invokeai/app/invocations/grounded_sam.py

```diff
@@ -0,0 +1,203 @@
+        """Use Grounding DINO to detect bounding boxes for a set of labels in an image."""
+        # TODO(ryand): I copied this "."-handling logic from the transformers example code. Test it and see if it
+        # actually makes a difference.
+        labels = [label if label.endswith(".") else label + "." for label in labels]
+
+        def load_grounding_dino(model_path: Path):
+            grounding_dino_pipeline = pipeline(
+                model=str(model_path),
+                task="zero-shot-object-detection",
+                local_files_only=True,
+                # TODO(ryand): Setting the torch_dtype here doesn't work. Investigate whether fp16 is supported by the
+                # model, and figure out how to make it work in the pipeline.
+                # torch_dtype=TorchDevice.choose_torch_dtype(),
+            )
+            assert isinstance(grounding_dino_pipeline, ZeroShotObjectDetectionPipeline)
+            return GroundingDinoPipeline(grounding_dino_pipeline)
+
+        with context.models.load_remote_model(source=GROUNDING_DINO_MODEL_ID, loader=load_grounding_dino) as detector:
+            assert isinstance(detector, GroundingDinoPipeline)
+            return detector.detect(image=image, candidate_labels=labels, threshold=threshold)
+
+    def _segment(
+        self,
+        context: InvocationContext,
+        image: Image.Image,
+        detection_results: list[DetectionResult],
+    ) -> list[DetectionResult]:
+        """Use Segment Anything (SAM) to generate masks given an image + a set of bounding boxes."""
+
+        def load_sam_model(model_path: Path):
+            sam_model = AutoModelForMaskGeneration.from_pretrained(
+                model_path,
+                local_files_only=True,
+                # TODO(ryand): Setting the torch_dtype here doesn't work. Investigate whether fp16 is supported by the
+                # model, and figure out how to make it work in the pipeline.
+                # torch_dtype=TorchDevice.choose_torch_dtype(),
+            )
+            assert isinstance(sam_model, SamModel)
+
+            sam_processor = AutoProcessor.from_pretrained(model_path, local_files_only=True)
+            assert isinstance(sam_processor, SamProcessor)
+            return SegmentAnythingModel(sam_model=sam_model, sam_processor=sam_processor)
+
+        with (
+            context.models.load_remote_model(source=SEGMENT_ANYTHING_MODEL_ID, loader=load_sam_model) as sam_pipeline,
+        ):
+            assert isinstance(sam_pipeline, SegmentAnythingModel)
+
+            masks = sam_pipeline.segment(image=image, detection_results=detection_results)
+
+        masks = self._to_numpy_masks(masks)
+        masks = self._apply_polygon_refinement(masks)
+
+        for detection_result, mask in zip(detection_results, masks, strict=False):
+            detection_result.mask = mask
+
+        return detection_results
+
+    def _to_numpy_masks(self, masks: torch.Tensor) -> list[npt.NDArray[np.uint8]]:
```

Comment:
Maybe chain this entire function rather than a multi line variable assignment. We can also get rid of the `int()` transmute because the boolean indexing will result in that anyway. The following should work I think. But need to test.

`return [(mask > 0).cpu().numpy().astype(np.uint8) for mask in masks.float().permute(0, 2, 3, 1).mean(-1)]`

Reply:
I tidied up this function a bit to do slightly more work on the GPU before copying to the CPU. I kept as separate variable assignments for readability. Take another look and let me know what you think.